scapy/contrib/tcpros.py

```diff
@@ -0,0 +1,732 @@
+# scapy.contrib.status = loads
+# scapy.contrib.name = tcpros
+
+import struct
+from scapy.compat import raw
+from scapy.fields import (
+    LEIntField,
+    StrLenField,
+    FieldLenField,
+    StrFixedLenField,
+    ByteField,
+)
+from scapy.layers.http import HTTP, HTTPRequest, HTTPResponse
+from scapy.packet import Packet, Raw, PacketListField
+
+
+class TCPROS(Packet):
+    """
+    TCPROS is a transport layer for ROS Messages and Services. It uses standard
+    TCP/IP sockets for transporting message data. Inbound connections are
+    received via a TCP Server Socket with a header containing message data type
+    and routing information.
+
+    This class focuses on capturing the ROS Slave API
+
+    An example package is presented below:
+
+        B0 00 00 00 26 00 00 00 63 61 6C 6C 65 72 69 64  ....&...callerid
+        3D 2F 72 6F 73 74 6F 70 69 63 5F 38 38 33 30 35  =/rostopic_88305
+        5F 31 35 39 31 35 33 38 37 38 37 35 30 31 0A 00  _1591538787501..
+        00 00 6C 61 74 63 68 69 6E 67 3D 31 27 00 00 00  ..latching=1'...
+        6D 64 35 73 75 6D 3D 39 39 32 63 65 38 61 31 36  md5sum=992ce8a16
+        38 37 63 65 63 38 63 38 62 64 38 38 33 65 63 37  87cec8c8bd883ec7
+        33 63 61 34 31 64 31 1F 00 00 00 6D 65 73 73 61  3ca41d1....messa
+        67 65 5F 64 65 66 69 6E 69 74 69 6F 6E 3D 73 74  ge_definition=st
+        72 69 6E 67 20 64 61 74 61 0A 0E 00 00 00 74 6F  ring data.....to
+        70 69 63 3D 2F 63 68 61 74 74 65 72 14 00 00 00  pic=/chatter....
+        74 79 70 65 3D 73 74 64 5F 6D 73 67 73 2F 53 74  type=std_msgs/St
+        72 69 6E 67                                      ring
+
+        Sources:
+            - http://wiki.ros.org/ROS/TCPROS
+            - http://wiki.ros.org/ROS/Connection%20Header
+            - https://docs.python.org/3/library/struct.html
+            - https://scapy.readthedocs.io/en/latest/build_dissect.html
+
+        TODO:
+            - Extend to support subscriber's interactions
+            - Unify with subscriber's header
+
+        NOTES:
+            - 4-byte length + [4-byte field length + field=value ]*
+            - All length fields are little-endian integers. Field names and
+                values are strings.
+            - Cooked as of ROS Melodic Morenia v1.14.5.
+    """
+
+    name = "TCPROS"
+
+    def guess_payload_class(self, payload):
+        string_payload = payload.decode("iso-8859-1")  # decode to string
+        # for search
+
+        # flag indicating if the TCPROS encoding format is met
+        #   4-byte length + [4-byte field length + field=value ]*
+        total_length = len(payload)
+        total_length_payload = struct.unpack("<I", payload[:4])[0]
+        remain = payload[4:]
+        remain_len = len(remain)
+        # flag of the encoding format
+        flag_encoding_format = (total_length > total_length_payload) and (
+            total_length_payload == remain_len
+        )
+
+        # # Debug
```

Comment:
You could use `if conf.debug_dissector:` if you want to preserve this debug code

Reply:
Thanks for the hint @polybassa, fixed at 92f21a48fd3a5c065c45b49a2900f8f6742f8947